keps/0029-operator-dependencies.md

```diff
@@ -0,0 +1,203 @@
+
+1. Build a [dependency graph](https://en.wikipedia.org/wiki/Dependency_graph) by transitively expanding top-level `deploy` plan using operator-tasks as vertices and their execution order (`a` needs `b` to be installed first) as edges
+2. Perform cycle detection and fail if circular dependencies found. We could additionally run this check on the client-side as part of the `kudo package verify` command to improve the UX
+3. If no cycles were found we traverse the dependency graph in the topological order (e.g. using the reversed post-order) and execute all vertices
+
+Let's take a look at an example. Here is a simplified operator `AA` with a few dependencies:
+
+```text
+AA
+├── BB
+│   ├── EE
+│   │   ├── H
+│   │   └── I
+│   ├── F
+│   └── GG
+│       ├── J
+│       └── K
+├── CC
+│   ├── L
+│   └── M
+└── D
+
+Legend:
+- Operators and operator-tasks are marked with double letters e.g. 'AA' or `BB`
+- other tasks are marked with single letters e.g. 'D'
+- direct children of an operator are the 'deploy' plan steps e.g. for 'AA' deploy steps are 'BB', 'CC' and 'D'
+```
+
+In the first step we build a dependency graph. A set of all graph vertices (which are task-operators) `S` is defined as `S = {AA, BB, CC, EE, GG}`. A transitive relationship `R` between the vertices is defined as `(a, b) ∈ S` meaning that _`a` needs `b` deployed first_. The transitive relationship for the above example is: `R = { (AA,BB), (AA,CC), (BB,EE), (BB,GG) }`. The resulting topological order `O` is therefor `O = (EE, GG, BB, CC, AA)` which has no cycles.
+
+The instance controller (IC) then traverses the dependency graph in the evaluation order `O` and executes each vertex. Practically this means creating a corresponding `Instance` resource (`Operator` and `OperatorVersion` already exist) and wait for it to become healthy, meaning that its `deploy` plan was executed successfully. KUDO manager already has [health check](https://github.com/kudobuilder/kudo/blob/master/pkg/engine/health/health.go#L78) for `Instance` resources implemented. We would additionally add the top-level `Instance` reference (e.g. `AA`) to the `ownerReferences` list of each dependency `Instance` (e.g. `BB`). This would help with determining which `Instance` belongs to which operator and additionally help us with [operator uninstalling](#uninstalling).
+
+The status of the execution can be seen as usual as part of the `Instance.Status`. We could additionally forward the status of a dependency `Instance` to the top-level `Instance.Status` to simplify the overview.
+
+##### Dependencies Parameterization
+
+First, we need to provide parameters to different operators separately. For individual parameters (`-p` option) we can use namespaced names e.g. `-p <operatorName>.<key>=<value>`. For the parameter files (`--parameterFile` option) we could extend `parameterFile` schema (which is currently a simple map) with top-level fields e.g.:
```

Comment:
Is that enough? I'm thinking about having multiple instances of the same Operator:
For example a top-level operator requires a) Zookeeper, and b) Kafka, which itself requires zookeeper as well.

In that case we might want to specify different amounts of nodes for each zookeeper, right? I guess we could also allow to use the same zk instance, but that would probably make things a lot more difficult?

Reply:
Additionally, in KEP-28 there is an idea to have plan-specific parameters which might be prefixed with the plan names, so this might clash there. 

Thinking about it, maybe it would work if we'd use the "task" name as a prefix, and make task-specific parameters instead of plan specific ones?

Reply:
>  I'm thinking about having multiple instances of the same Operator:
For example a top-level operator requires a) Zookeeper, and b) Kafka, which itself requires zookeeper as well.

I've been thinking about it - what does it actually mean, to depend on **an already existing* `Instance`?  

a) **resource referencing:** for an existing Zookeeper instance `zook`, you'd most probably like to reference it in e.g. Kafkas `ZOOKEEPER_URI: zook-zookeeper-0.zook-hs:2181`. Since we don't support terraform style `Instance` resource referencing you could simply "hardcode" it.
b) **installation check:** make sure that the `zook` instance is running healthy. Not yet supported.

But otherwise, you're right - reusing the same `Instance` during deployment is not supported as it would create circular dependency. I will add this to the KEP.

Reply:
> Additionally, in KEP-28 there is an idea to have plan-specific parameters which might be prefixed with the plan names, so this might clash there.

I hope we would go with proposal nr. 5 and avoid _"plan-specific parameters"_